File: src/main/java/endytkn/randomEvents/events/ZombieFightEvent.java
package endytkn.randomEvents.events;

import endytkn.randomEvents.baseEvents.groupFight.GroupFight;
import endytkn.randomEvents.baseEvents.groupFight.GroupFightBaseEvent;
import endytkn.randomEvents.randomEvent.RandomEvent;
import endytkn.randomEvents.randomEvent.RandomEventsCategory;
import endytkn.randomEvents.randomEvent.RandomEventsRarity;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.monster.Zombie;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.enchantment.Enchantments;

import java.util.*;

public class ZombieFightEvent extends GroupFightBaseEvent {
    public ZombieFightEvent() {
        super(true);
        categories = List.of(RandomEventsCategory.GROUP_FIGHT, RandomEventsCategory.DAY_NIGHT, RandomEventsCategory.WEATHER_CLEAR, RandomEventsCategory.WEATHER_THUNDER, RandomEventsCategory.WEATHER_RAIN);
        this.rarity = RandomEventsRarity.COMMON;
        this.eventTag = "zombie_fight";
    }

    @Override
    public RandomEvent create() {
        return new ZombieFightEvent();
    }

    @Override
    public void onPrepare() {
        int minionsCount = new Random().nextInt(3 + this.playersGroup.size() * 2, 5 + this.playersGroup.size() * 2);
        int leaderCount = minionsCount / 3;
        Map<UUID, Mob> zombieMobs = new HashMap<>();

        for (int i = 0; i < minionsCount; i++) {
            spawnZombie(zombieMobs, false);
        }

        for (int i = 0; i < leaderCount; i++) {
            spawnZombie(zombieMobs, true);
        }

        // Create the group fight
        GroupFight zombieGroup = new GroupFight(this, "zombie", zombieMobs, null);
        mobGroupies.put(zombieGroup.groupName, zombieGroup);
        super.onPrepare();
    }

    private void spawnZombie(Map<UUID, Mob> zombieMobs, boolean isLeader) {
        Zombie mob = EntityType.ZOMBIE.create(level);

        if (mob != null) {
            if (isLeader) {
                this.addEntityToEvent(mob);
                mob.setHealth(mob.getHealth() * 2); // Double health for leaders
                mob.setCustomName(Component.literal("Zombie Leader"));
                ItemStack helmet = new ItemStack(Items.GOLDEN_HELMET);
                helmet.enchant(Enchantments.UNBREAKING, 5);
                mob.setItemSlot(EquipmentSlot.HEAD, helmet);
            }

            mob.setCanPickUpLoot(false);
            zombieMobs.put(mob.getUUID(), mob);
        }
    }
}

File: src/main/kotlin/endytkn/randomEvents/events/ZombieFightEvent.kt
package endytkn.randomEvents.events

import endytkn.randomEvents.baseEvents.GroupFight.GroupFight
import endytkn.randomEvents.baseEvents.GroupFight.GroupFightBaseEvent
import endytkn.randomEvents.randomEvents.RandomEvent
import endytkn.randomEvents.randomEvents.RandomEventsCategories
import endytkn.randomEvents.randomEvents.RandomEventsRarity
import net.minecraft.network.chat.Component
import net.minecraft.world.InteractionHand
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.effect.MobEffects
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.EquipmentSlot
import net.minecraft.world.entity.Mob
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.enchantment.Enchantments
import java.util.UUID
import kotlin.random.Random

class ZombieFightEvent : GroupFightBaseEvent(true) {
    companion object {
        val name: String = "skeleton_fight"
    }

    var minionsCount: Int? = null
    var leaderCount: Int? = null

    init {
        category = RandomEventsCategories.GROUP_FIGHT
        rarity = RandomEventsRarity.COMMON
        this.eventTag = "zombieFight"
    }

    override fun create(): RandomEvent {
        return ZombieFightEvent()
    }

    override fun onPrepare() {
        minionsCount = Random.nextInt(3 + this.playersGroup!!.size * 2, 5 + this.playersGroup!!.size * 2)
        leaderCount = minionsCount!! / 3
        val zombieMobs = mutableMapOf<UUID, Mob>()

        repeat(minionsCount!!) { index ->
            var mobType: EntityType<out Mob> = EntityType.ZOMBIE
            var mob = mobType.create(level!!)
            mob?.let {
                it.setCanPickUpLoot((false))
                zombieMobs.put(it.uuid, it)
            }
        }

        repeat(leaderCount!!) { index ->
            var mobType: EntityType<out Mob> = EntityType.ZOMBIE
            var mob = mobType.create(level!!)

            mob?.let {
                val weapon =  ItemStack(Items.GOLDEN_SWORD)
                it.health = (mob.health * 2)
                it.customName = Component.literal("Zombie Leader")
                it.armorSlots
                val helmet = ItemStack(Items.GOLDEN_HELMET)
                helmet.enchant(Enchantments.UNBREAKING, 5)
                it.setItemSlot(EquipmentSlot.HEAD, helmet)
                it.setCanPickUpLoot((false))
                it.setItemInHand(InteractionHand.MAIN_HAND, weapon)
                zombieMobs.put(it.uuid, it)
            }
        }
        val zombieGroup = GroupFight(this, "zombie", zombieMobs, null)
        mobGroupies.put(zombieGroup.groupName, zombieGroup)
        super.onPrepare()
    }
}